db.go

```diff
@@ -398,198 +397,196 @@ func (s *ColumnStore) DB(ctx context.Context, name string) (*DB, error) {
-	level.Info(db.logger).Log("msg", "replaying WAL")
-	start := time.Now()
-	// persistedTables is a map from a table name to the last transaction
-	// persisted.
+func (db *DB) openWAL(ctx context.Context, firstIndex uint64) (WAL, error) {
 	persistedTables := map[string]uint64{}
-	if err := db.wal.Replay(firstIndex, func(tx uint64, record *walpb.Record) error {
-		switch e := record.Entry.EntryType.(type) {
-		case *walpb.Entry_TableBlockPersisted_:
-			persistedTables[e.TableBlockPersisted.TableName] = tx
-		}
-		return nil
-	}); err != nil {
-		return fmt.Errorf("first WAL replay: %w", err)
-	}
-
 	lastTx := uint64(0)
-	if err := db.wal.Replay(firstIndex, func(tx uint64, record *walpb.Record) error {
-		lastTx = tx
-		switch e := record.Entry.EntryType.(type) {
-		case *walpb.Entry_NewTableBlock_:
-			entry := e.NewTableBlock
-			var schema proto.Message
-			switch v := entry.Schema.(type) {
-			case *walpb.Entry_NewTableBlock_DeprecatedSchema:
-				schema = v.DeprecatedSchema
-			case *walpb.Entry_NewTableBlock_SchemaV2:
-				schema = v.SchemaV2
-			default:
-				return fmt.Errorf("unhandled schema type: %T", v)
-			}
-
-			var id ulid.ULID
-			if err := id.UnmarshalBinary(entry.BlockId); err != nil {
-				return err
-			}
 
-			if lastPersistedTx, ok := persistedTables[entry.TableName]; ok && tx < lastPersistedTx {
-				// This block has already been successfully persisted, so we can
-				// skip it.
-				return nil
-			}
-
-			tableName := entry.TableName
-			table, err := db.GetTable(tableName)
-			var tableErr ErrTableNotFound
-			if errors.As(err, &tableErr) {
-				schema, err := dynparquet.SchemaFromDefinition(schema)
-				if err != nil {
-					return fmt.Errorf("initialize schema: %w", err)
-				}
-				config := NewTableConfig(schema)
-				// TODO: May be we should acquire mutux lock when mutating d.roTables and d.tables?
-				// s.DB creates read only Table instance when BucketStore is configured.
-				// Make sure to use the existing Table instace instead of creating new one to avoid dangling instance.
-				table, ok := db.roTables[tableName]
-				if ok {
-					delete(db.roTables, tableName)
-					table.config = config
-				} else {
-					table, err = newTable(
-						db,
-						tableName,
-						config,
-						db.reg,
-						db.logger,
-						db.tracer,
-						db.wal,
-					)
-				}
-				if err != nil {
-					return fmt.Errorf("instantiate table: %w", err)
-				}
-
-				table.active, err = newTableBlock(table, 0, tx, id)
-				if err != nil {
-					return err
+	wal, err := wal.Open(
+		db.logger,
+		db.reg,
+		db.walDir(),
+		wal.WithFirstIndex(firstIndex),
+		// NOTE: 2 is passed here because we require 2 passes through the WAL.
+		// Once to establish all the storage blocks that have rotated out, and a second pass to replay all the writes.
+		// This prevents us from replaying writes that have already been persisted/dropped by a block rotation.
+		wal.WithReplayFunc(2, func(i int, tx uint64, record *walpb.Record) error {
```

Comment:
I find this interface a bit awkward. I think it should be enough for our purposes to call `wal.Open` first and then `wal.Replay` below.